File: lib/services/deepseek_client.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

import '../config/deepseek_config.dart';
import '../config/prophecy_style.dart';
import '../models/sensor_data.dart';
import '../utils/prophecy_prompt_builder.dart';

class DeepSeekException implements Exception {
  DeepSeekException(this.message, {this.statusCode});

  final String message;
  final int? statusCode;

  @override
  String toString() => message;
}

/// DeepSeek Chat Completions 客户端
class DeepSeekClient {
  DeepSeekClient({http.Client? httpClient}) : _http = httpClient ?? http.Client();

  final http.Client _http;

  /// 测试用：绕过真实 HTTP
  @visibleForTesting
  Future<http.Response> Function(
    Uri uri,
    Map<String, String> headers,
    String body,
  )? postOverride;

  Future<String?> generate({
    required String apiKey,
    required SensorData sensor,
    required int nonce,
    double temperature = ProphecyStyle.temperature,
  }) async {
    final key = apiKey.trim();
    if (key.isEmpty) return null;

    final chat = ProphecyPromptBuilder.buildMlxChat(sensor, nonce: nonce);
    const apiBase = 'https://api.deepseek.com';
    const chatPath = '/chat/completions';
    const timeoutSeconds = 30;
    final uri = Uri.parse('$apiBase$chatPath');
    final body = jsonEncode({
      'model': DeepSeekConfig.model,
      'messages': [
        {'role': 'system', 'content': chat.system},
        {'role': 'user', 'content': chat.user},
      ],
      'temperature': temperature,
      'max_tokens': ProphecyStyle.maxTokens,
      'stream': false,
    });
    final headers = {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $key',
    };

    try {
      final response = postOverride != null
          ? await postOverride!(uri, headers, body)
          : await _http
              .post(uri, headers: headers, body: body)
              .timeout(const Duration(seconds: timeoutSeconds));

      if (response.statusCode == 401 || response.statusCode == 403) {
        throw DeepSeekException('API 密钥无效或已过期', statusCode: response.statusCode);
      }
      if (response.statusCode < 200 || response.statusCode >= 300) {
        throw DeepSeekException(
          'DeepSeek 请求失败（${response.statusCode}）',
          statusCode: response.statusCode,
        );
      }

      final decoded =
          jsonDecode(utf8.decode(response.bodyBytes)) as Map<String, dynamic>;
      final choices = decoded['choices'] as List<dynamic>?;
      if (choices == null || choices.isEmpty) return null;

      final message = choices.first as Map<String, dynamic>;
      final content = (message['message'] as Map<String, dynamic>?)?['content'];
      if (content is! String || content.trim().isEmpty) return null;
      return content.trim();
    } on DeepSeekException {
      rethrow;
    } catch (e) {
      debugPrint('DeepSeek generate failed: $e');
      throw DeepSeekException('网络异常，请检查连接后重试');
    }
  }
}

File: server/app/services/deepseek_client.py
"""DeepSeek Chat Completions 客户端（移植自 lib/services/deepseek_client.dart）。"""

from __future__ import annotations

import httpx

from app.config import Settings
from app.models.sensor import SensorPayload
from app.services.prophecy_prompt_builder import build_mlx_chat

TEMPERATURE = 0.85
RETRY_TEMPERATURE = 0.7
MAX_TOKENS = 64


class DeepSeekError(Exception):
    def __init__(self, message: str, *, status_code: int | None = None) -> None:
        super().__init__(message)
        self.message = message
        self.status_code = status_code


class DeepSeekClient:
    def __init__(self, settings: Settings) -> None:
        self._settings = settings

    async def generate(
        self,
        sensor: SensorPayload,
        *,
        nonce: int,
        temperature: float = TEMPERATURE,
    ) -> str | None:
        api_key = self._settings.deepseek_api_key
        if not api_key:
            raise DeepSeekError("服务端未配置 DEEPSEEK_API_KEY")

        chat = build_mlx_chat(sensor, nonce=nonce)
        url = f"{self._settings.deepseek_api_base}/chat/completions"
        payload = {
            "model": self._settings.deepseek_model,
            "messages": [
                {"role": "system", "content": chat.system},
                {"role": "user", "content": chat.user},
            ],
            "temperature": temperature,
            "max_tokens": MAX_TOKENS,
            "stream": False,
        }
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
        }

        try:
            async with httpx.AsyncClient(
                timeout=self._settings.deepseek_timeout_seconds
            ) as client:
                response = await client.post(url, json=payload, headers=headers)
        except httpx.TimeoutException as exc:
            raise DeepSeekError("DeepSeek 请求超时") from exc
        except httpx.HTTPError as exc:
            raise DeepSeekError("网络异常，请检查连接后重试") from exc

        if response.status_code in (401, 403):
            raise DeepSeekError("API 密钥无效或已过期", status_code=response.status_code)
        if response.status_code < 200 or response.status_code >= 300:
            raise DeepSeekError(
                f"DeepSeek 请求失败（{response.status_code}）",
                status_code=response.status_code,
            )

        decoded = response.json()
        choices = decoded.get("choices") or []
        if not choices:
            return None

        message = choices[0].get("message") or {}
        content = message.get("content")
        if not isinstance(content, str) or not content.strip():
            return None
        return content.strip()
